File: src/StreamDb/Internal/Support/CRC32.cs
using JetBrains.Annotations;

namespace StreamDb.Internal.Support
{
    public static class Crc32
    {
        public const uint DefaultPolynomial = 0xedb88320;
        public const uint DefaultSeed = 0xffffffff;

        [NotNull]private static readonly uint[] defaultTable;

        static Crc32()
        {
            var createTable = new uint[256];
            for (int i = 0; i < 256; i++)
            {
                var entry = (uint)i;
                for (int j = 0; j < 8; j++) entry = (entry & 1) == 1 ? (entry >> 1) ^ DefaultPolynomial : entry >> 1;
                createTable[i] = entry;
            }

            defaultTable = createTable;
        }

        /// <summary>
        /// Compute the CRC for 
        /// </summary>
        /// <param name="buffer"></param>
        /// <returns></returns>
        public static uint Compute(byte[] buffer)
        {
            if (buffer == null) return 0;
            var crc = DefaultSeed;
            foreach (var b in buffer)
            {
                crc = (crc >> 8) ^ defaultTable[b ^ (crc & 0xff)];
            }
            return ~crc;
        }
    }
}

File: golang/internal/support/crc32.go
package support

const (
	DefaultPolynomial uint32 = 0xedb88320
	DefaultSeed uint32 = 0xffffffff
)

var (
	defaultTable []uint32
)

func init() {
	defaultTable = make([]uint32, 256)
	for i := 0; i < 256; i++ {
		entry := uint32(i)
		for j := 0; j < 8; j++ {
			if (entry & 1) == 1 {
				entry = (entry >> 1) ^ DefaultPolynomial
			} else {
				entry = entry >> 1
			}
		}
		defaultTable[i] = entry
	}
}

func ComputeCRC32(buffer []byte) uint32 {
	if buffer == nil {
		return 0
	}
	var crc = DefaultSeed
	for i := 0; i < len(buffer); i++ {
		b := uint32(buffer[i])
		crc = (crc >> 8) ^ defaultTable[b^(crc&0xff)]
	}
	return ^crc
}